keps/sig-network/1880-services-ip-ranges/README.md

```diff
@@ -0,0 +1,1030 @@
+from there. To mititage this problem, we'll set a grace period of 60 seconds on the ServiceIPrange controller
+to remove the finalizer, if an IP address is created during this time we'll be able to block the deletion
+and inform the user.
+
+For any Service and IPAddress that doesn't belong to a ServiceIPRange the controller will raise an event
+informing the user, keeping the same behavior
+
+```go
+// cluster IP is out of range
+c.recorder.Eventf(&svc, nil, v1.EventTypeWarning, "ClusterIPOutOfRange", "ClusterIPAllocation", "Cluster IP [%v]:%s is not within the service CIDR %s; please recreate service",
+```
+
+- Valid Service and ServiceIPRange but not IPAddress
+
+It can happen that an user forcefully delete an IPAddress, in this case, the controller will regenerate the IPAddress.
+
+During this time, there is a chance that an apiserver tries to allocate this IPAddress, with a possible situation where
+2 Services has the same IPAddress. In order to avoid it, the Allocator will not delete an IP from its local cache
+until it verifies that the Service associated has been deleted too.
+
+- Valid IPAddress and ServiceIPRange but no Service
+
+The IPAddress will be deleted and an event generated if the controller determines that the IPAddress
+is orphan [(see Allocator section)](#Allocator)
+
+#### Resizing Service IP Ranges
+
+One of the most common problems users may have is how to scale up or scale down their service IP ranges.
+
+Let's see an example on how these operations will be implemented.
+
+Assume we have  configured a service range 10.0.0.0/24 and its fully used, we can:
+
+1. Add another /24 range 10.0.1.0/24 and keep the previous one
+2. Add an overlapping larger range 10.0.0.0/23
+
+The user can now remove the first /24 range, since the new range covers all the existing IP Addresses
+
+The same applies for scaling down, meanwhile the new range contains all the existing IPAddresses, the old
+range will be safely deleted.
+
+However, there can be a race condition during the operation of scaling down, since the propagation of the
+deletion can take some time, one allocator can successfully allocate an IP address out of new range (but still
+inside of the old range).
+
+There is one controller that will periodically check that 1on1 relation between IPAddresses and Services is
+correct, and will start sending events to warn the user that it has to fix/recreate the corresponding Service.
+
+
+
+#### API
+
+
+```go
+// ServiceIPRange defines a range of IPs using CIDR format (192.168.0.0/24 or 2001:db2::0/64).
+type ServiceIPRange struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ServiceIPRangeSpec   `json:"spec,omitempty"`
+}
+
+
+// ServiceIPRangeSpec describe how the ServiceIPRange's specification looks like.
+type ServiceIPRangeSpec struct {
+	// CIDR is an IP block in CIDR notation "10.0.0.0/8" or "fd12:3456:789a:1::/64"
+  // in using its canonical representation (rfc5952)
```

Comment:
you say canonical but the example you give isnt' :)

I don't think it needs to  be canonical, though?

Reply:
if we move to single-stack what if we make this the name instead of a spec field, this will handle duplications

Reply:
We can't make it the name because we need to allow overlaps in the defaults.  E.g. apiserver makes a default, user deletes it, apiserver makes a new default with the same range.

Reply:
then canonical seem superfluos

Reply:
That was my argument :)